readthedocs/search/serializers.py

```diff
@@ -0,0 +1,143 @@
+"""
+Serializers for the ES's search result object.
+
+.. note::
+   Some fields are re-named to make their meaning more clear.
+   They should be renamed in the ES index too.
+"""
+
+import itertools
+from operator import attrgetter
+
+from django.shortcuts import get_object_or_404
+from rest_framework import serializers
+
+from readthedocs.core.resolver import resolve
+from readthedocs.projects.models import Project
+
+
+class ProjectHighlightSerializer(serializers.Serializer):
+
+    name = serializers.ListField(child=serializers.CharField(), default=list)
+    slug = serializers.ListField(child=serializers.CharField(), default=list)
+    description = serializers.ListField(child=serializers.CharField(), default=list)
+
+
+class ProjectSearchSerializer(serializers.Serializer):
+
+    type = serializers.CharField(default='project', source=None, read_only=True)
+    name = serializers.CharField()
+    slug = serializers.CharField()
+    link = serializers.CharField(source='url')
+    highlight = ProjectHighlightSerializer(source='meta.highlight', default=dict)
+
+
+class PageHighlightSerializer(serializers.Serializer):
+
+    title = serializers.ListField(child=serializers.CharField(), default=list)
+
+
+class PageSearchSerializer(serializers.Serializer):
+
+    type = serializers.CharField(default='page', source=None, read_only=True)
+    project = serializers.CharField()
+    version = serializers.CharField()
+    title = serializers.CharField()
+    path = serializers.CharField(source='full_path')
+    link = serializers.SerializerMethodField()
+    highlight = PageHighlightSerializer(source='meta.highlight', default=dict)
+    inner_hits = serializers.SerializerMethodField()
+
+    def get_link(self, obj):
+        # TODO: optimize this to not query the db for each result.
+        project = Project.objects.filter(slug=obj.project).first()
+        if project:
+            return resolve(
```

Comment:
We talked about having the search API return relative results on the in-doc search (`/en/latest/foo.html` instead of `https://docs.example.com/en/latest/foo.html`), so that we can support custom domains and subpaths with fewer queries. It might be worth adding some logic here to support that.

Reply:
If I'm able to implement this solution https://github.com/readthedocs/readthedocs.org/pull/7157/files#r440502564 we won't need to do that to save queries

Reply:
I think it's better to return the relative URL in docs pages anyway -- it's what sphinx does, and it stops us from hitting issues with domain canonicalization in the responses (eg. pip.rtd.io returning links to pip.pypa.io). It's just better in general for us to not guess what domain to use, unless we have to.

Reply:
that makes sense, but we still need to generate the full link for results in the dashboard

Reply:
Yep 👍